com.unity.shadergraph/Editor/Data/Nodes/Input/Basic/ColorRampNode.cs

```diff
@@ -0,0 +1,213 @@
+                    );
+                }
+
+                for (var i = 1; i < alphaKeys.Length; i++)
+                {
+                    sb.AppendLine(
+                        "{0}.a = lerp({0}.a, {1}, saturate(({2} - {3}) / ({4} - {3})));",
+                        outName, alphaKeys[i].alpha, inValue,
+                        alphaKeys[i - 1].time, alphaKeys[i].time
+                    );
+                }
+
+                sb.AppendLine("#ifndef UNITY_COLORSPACE_GAMMA");
+                sb.AppendLine("{0}.rgb = SRGBToLinear({0}.rgb);", outName);
+                sb.AppendLine("#endif");
+            }
+            else
+            {
+                // Preview mode: Call the ramp function with uniforms.
+                sb.AppendLine(
+                    "{0}_color({1}, {2}.rgb, {3}_c0, {3}_c1, {3}_c2, {3}_c3, {3}_c4, {3}_c5, {3}_c6, {3}_c7);",
+                    GetFunctionName(), inValue, outName, GetVariableNameForNode()
+                );
+
+                sb.AppendLine(
+                    "{0}_alpha({1}, {2}.a, {3}_a0, {3}_a1, {3}_a2, {3}_a3, {3}_a4, {3}_a5, {3}_a6, {3}_a7);",
+                    GetFunctionName(), inValue, outName, GetVariableNameForNode()
+                );
+            }
+
+            visitor.AddShaderChunk(sb.ToString(), false);
+        }
+
+        public override void CollectShaderProperties(PropertyCollector properties, GenerationMode generationMode)
+        {
+            base.CollectShaderProperties(properties, generationMode);
+
+            if (generationMode != GenerationMode.Preview) return;
+
+            var colorKeys = m_Ramp.colorKeys;
+            var alphaKeys = m_Ramp.alphaKeys;
+            var name = GetVariableNameForNode();
+
+            for (var i = 0; i < 8; i++)
+            {
+                properties.AddShaderProperty(new Vector4ShaderProperty()
+                {
+                    overrideReferenceName = name + "_c" + i,
+                    generatePropertyBlock = false,
+                    value = KeyToVector(colorKeys[Mathf.Min(i, colorKeys.Length - 1)])
+                });
+
+                properties.AddShaderProperty(new Vector2ShaderProperty()
+                {
+                    overrideReferenceName = name + "_a" + i,
+                    generatePropertyBlock = false,
+                    value = KeyToVector(alphaKeys[Mathf.Min(i, alphaKeys.Length - 1)])
+                });
+            }
+        }
+
+        public override void CollectPreviewMaterialProperties(List<PreviewProperty> properties)
+        {
+            base.CollectPreviewMaterialProperties(properties);
+
+            var colorKeys = m_Ramp.colorKeys;
+            var alphaKeys = m_Ramp.alphaKeys;
+            var name = GetVariableNameForNode();
+
+            for (var i = 0; i < 8; i++)
+            {
+                properties.Add(new PreviewProperty(PropertyType.Vector4)
+                {
+                    name = name + "_c" + i,
```

Comment:
This should be pre-generated to avoid allocation, as it is called every frame during preview rendering.